netty/src/main/java/io/grpc/netty/WriteBufferingAndExceptionHandler.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.netty;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.Status;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.ReferenceCountUtil;
+import java.net.SocketAddress;
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * Buffers all writes until either {@link #writeBufferedAndRemove(ChannelHandlerContext)} or
+ * {@link #fail(ChannelHandlerContext, Throwable)} is called. This handler allows us to
+ * write to a {@link io.netty.channel.Channel} before we are allowed to write to it officially
+ * i.e.  before it's active or the TLS Handshake is complete.
+ */
+final class WriteBufferingAndExceptionHandler extends ChannelDuplexHandler {
+
+  private final Queue<ChannelWrite> bufferedWrites = new ArrayDeque<>();
+  private final ChannelHandler next;
+  private boolean writing;
+  private boolean flushRequested;
+  private Throwable failCause;
+
+  WriteBufferingAndExceptionHandler(ChannelHandler next) {
+    this.next = checkNotNull(next, "next");
+  }
+
+  @Override
+  public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+    ctx.pipeline().addBefore(ctx.name(), null, next);
+    super.handlerAdded(ctx);
+  }
+
+  /**
+   * If this channel becomes inactive, then notify all buffered writes that we failed.
+   */
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) {
+    Status status = Status.UNAVAILABLE.withDescription(
+        "Connection closed while performing protocol negotiation for " + ctx.pipeline().first());
```

Comment:
I made it show the first two, since I expect the handler(s) nearest the network to be doing the negotiation (though they may not be instanceof ProtocolNegotiator.Handler).    WDYT?

Reply:
Hmmm... I can live with two, although again, I think the end of the pipeline is a bit more interesting. I'd be willing to include all of `names()` in the message :) Whatever.